Feature/modularization core

Micpol opened the pull request:
**Breaking changes**

GDPR related classes have been moved from rodo to gdpr package
Cross_ads promotedApp classes are moved from library module to crossads module
BitmapLoader, Utils are now in utils Package

**Fixes**
rewardedAd too many calls to load an ad

**Use as a local dependency**
From now in order to use api_android modules as inside local dependencies you should copy buildSrc folder from api_android project to your project root directory. 
Because of that, when testing new libraries or dependencies against api_android you should always updates both api_android and your root project buildSrc/Dependencies.kt files.

Core module still needs a bit work on delay between interstitial ads inside apps.

rogalz committed:
Add Xiaomi Mi A1 (Kuba) test device

rogalz committed:
If interstitial is not showed (because delay) call not showed callback

Micpol committed:
Adds onDestroy to rewardedAd

Micpol committed:
Removes unnecessary null checks

WrobelKrakow committed:
Merge pull request #72 from netigenkluzowicz/fix-interstitial-delay

Fix interstitial delay

Micpol committed:
Fixes imports, adds basic stuff to NetigenSplashFragment

Micpol committed:
Updates repositories

Micpol committed:
Deletes buildSrc directory

Micpol committed:
Removes unnecessary tasks

Micpol committed:
Adds dependencies and versions as ext to main build.gradle, adds proper config values

Micpol committed:
Reverts gradle files version before merging fix/update-versions

Micpol committed:
Configure project to use ext versions and dependencies

Micpol committed:
Attempt to make buildSrc and kotlin-dsl dependencies work

Micpol committed:
Changes hardcoded values to one fetched from Dependencies.kt

Micpol committed:
Deletes unnecessary repositories block, applies android block from android.gradle

Micpol committed:
Reverts android.gradle to state with hardcoded values

Micpol committed:
Deletes android.gradle file, reintroduces android { } block to every module again

Micpol committed:
Deletes unneeded libraries and repositories.gradle file

Micpol committed:
Updates library version, removes recyclerView version

WrobelKrakow committed:
Merge pull request #73 from netigenkluzowicz/fix/empty-classes

Fix/empty classes

Micpol committed:
Fixes jitpack build containing empty jars

Micpol committed:
Adds languages package

Micpol committed:
Moves initialisation to init block

Micpol committed:
Adds splash screen logic, refactors code

Micpol committed:
Implements basic onAttach

Micpol committed:
Adds methods to show/hide bannerAd divider, removes splash screen logic

Micpol committed:
Deletes unnecessary repositories

Micpol committed:
Adds splash logic

Micpol committed:
InterstitialAdsManager and splash changes
- SplashLoader now takes Unit as parameter,
- AdsManager exposes interstitialAdManager method to launchSplashLoader
-GDPRDialogFragment is now accessed with elvis instead of !! operator
-Reformats code

Micpol committed:
GDPR dialog fragment now calls dismiss by itself

Micpol committed:
Changes shouldShowInterstitial to isDesignedForFamily

Micpol committed:
Changes constRodo name to ConstGDPR and converts to Kotlin

Micpol committed:
Implements proper designed for family logic

Micpol committed:
Changes netigenMainActivity reference to nullable

Micpol committed:
Adds payments as module dependency on core
-onItemNotBought will now properly call onNoAdsNotBought

Micpol committed:
Removes unnecessary type check

Micpol committed:
Refactors code, adds testRewardedAd id, Config is now initialized by lazy

Micpol committed:
Fixes interstitialAd not showing on splash, adds proper delay between ads

Micpol committed:
Fixes initAds not called in designForFamily apps

Micpol committed:
Converts Config and builder to Kotlin

Micpol committed:
Removes unnecessary test device, refactors code

Micpol committed:
Fixes a bug where app couldn't load another fragment when ad wasn't ready

Micpol committed:
Adapts viewModel to new Config class

Micpol committed:
Adds payments onDestroy call in NetigenActivity onDestroy

Micpol committed:
Adds test sku in debug builds

Micpol committed:
Fixes ads shown on noAdsBought

Micpol committed:
Reformats code, optimizes imports

Micpol committed:
Fixes problem with checkIfSamsung returning always true

WrobelKrakow commented on core/src/main/java/pl/netigen/core/ads/AdsManager.kt:
```diff
@@ -125,11 +124,24 @@ class AdsManager(var viewModel: NetigenViewModel, val activity: AppCompatActivit
+    fun splashScreenOnDestroy() {
+        interstitialAdManager.interstitialAdHandler.removeCallbacksAndMessages(null)
+    }
+
```
There is not method for removing rewardAdsListeners. Maybe it is not required

Micpol replied:
I didn't find a use for it in Notepad App for I use it only in one place, though there might be some specific cases when you'd like to remove them, for example having rewardedAds in multiple fragments, without removing callbacks it could cause some memory leaks.

WrobelKrakow commented on core/src/main/java/pl/netigen/core/ads/InterstitialAdManager.kt:
```diff
@@ -14,11 +14,11 @@ class InterstitialAdManager(private val viewModel: NetigenViewModel, val activit
     private val DEFAULT_MIN_WAIT: Long = 2000
     private val DEFAULT_MAX_WAIT: Long = 7000
```
Rename this vals or move to top level

Micpol replied:
Done

WrobelKrakow commented on core/src/main/java/pl/netigen/core/ads/InterstitialAdManager.kt:
```diff
@@ -28,6 +28,7 @@ class InterstitialAdManager(private val viewModel: NetigenViewModel, val activit
+        load(activity)
     }
 
     fun loadIfPossible() {
```
Strange logic to check,, load only when interstitialAdError is true

Micpol replied:
Yup, forgot to refactor it while replacing with old AdmobManager.

WrobelKrakow commented on core/src/main/java/pl/netigen/core/ads/InterstitialAdManager.kt:
```diff
@@ -36,132 +37,136 @@ class InterstitialAdManager(private val viewModel: NetigenViewModel, val activit
+        }
     }
 
     private fun onLoaded(showInterstitialListener: ShowInterstitialListener) {
```
We must review this logic

Micpol replied:
Resolved with https://github.com/netigenkluzowicz/api_android/pull/74/commits/a3f7686ae18a6633410e6a5162aacd9c97881a7e

WrobelKrakow commented on core/src/main/java/pl/netigen/core/ads/RewardedAd.kt:
```diff
@@ -30,7 +29,7 @@ class RewardedAd(var viewModel: NetigenViewModel, val activity: AppCompatActivit
     private lateinit var rewardedVideoAd: RewardedVideoAd
```
unnecessary lateinit

Micpol replied:
Done, initialization moved to init block

WrobelKrakow commented on core/src/main/java/pl/netigen/core/ads/RewardedAd.kt:
```diff
@@ -57,20 +56,22 @@ class RewardedAd(var viewModel: NetigenViewModel, val activity: AppCompatActivit
         }
     }
 
+    fun onDestroy() {
```
Maybe remove listener too

WrobelKrakow commented on core/src/main/java/pl/netigen/core/config/Config.kt:
```diff
@@ -1,35 +1,30 @@
+class Config constructor(val inDebugMode: Boolean, val isSamsung: Boolean,
+                         var isNoAdsBought: Boolean = false, var isNoAdsPaymentAvailable: Boolean = true,
+                         var isDesignedForFamily: Boolean = false, var rewardedAdId: String? = null,
+                         var adMobAppId: String? = null, var publishersId: Array<String> = arrayOf("pub-4699516034931013"),
```
unused adMobAppId

Micpol replied:
Right, also appId should now be set in AndroidManifest

WrobelKrakow commented on core/src/main/java/pl/netigen/core/netigenapi/NetigenFragment.kt:
```diff
@@ -1,4 +1,17 @@
+
+abstract class NetigenFragment : Fragment() {
+
+    var netigenMainActivity: NetigenMainActivity<NetigenViewModel>? = null
```
Why not nonnull lateinit?

Micpol replied:
I think this class is not needed actually, for it introduces no real profit. Deleting it

WrobelKrakow commented on core/src/main/java/pl/netigen/core/netigenapi/NetigenMainActivity.kt:
```diff
@@ -2,104 +2,136 @@ package pl.netigen.core.netigenapi
-                adsManager?.loadRewardedVideo()
+    private fun observeNoAds() {
+        viewModel.noAdsLiveData.observe(this, Observer {
+            if (it) {
```
why not call onNoAdsBought()?

Micpol replied:
Fixed with https://github.com/netigenkluzowicz/api_android/pull/74/commits/00c6bf8aa1d8a1d7c1e766d2b3e7f3c297c81308

WrobelKrakow commented on core/src/main/java/pl/netigen/core/netigenapi/NetigenSplashFragment.kt:
```diff
@@ -1,4 +1,162 @@
+abstract class NetigenSplashFragment<ViewModel : NetigenViewModel> : Fragment(), GDPRDialogFragment.GDPRClickListener {
+
+    open lateinit var viewModel: ViewModel
+    var netigenMainActivity: NetigenMainActivity<NetigenViewModel>? = null
+    private var gdprDialogFragment: GDPRDialogFragment? = null
```
Why nullable Types?

Micpol replied:
For the same reason getActivity returns a nullable type. And GDPRDialogFragment will be initialized only when ConsentStatus is Unknown, there's no need to initialize it in every other situation

WrobelKrakow commented on core/src/main/java/pl/netigen/core/netigenapi/NetigenSplashFragment.kt:
```diff
@@ -1,4 +1,162 @@
+        super.onDestroyView()
+    }
+
+    fun onNoAdsPaymentProcessingFinished(noAdsBought: Boolean) {
```
unused

Micpol replied:
Done, removed

WrobelKrakow commented on core/src/main/java/pl/netigen/core/netigenapi/NetigenViewModel.kt:
```diff
@@ -1,38 +1,50 @@
-    abstract fun prepareConfigBuilder(): ConfigBuilder
+    var isSplashInBackground: Boolean = false
+    abstract fun prepareConfigBuilder(): Config
+    open val config: Config by lazy {
```
remove open

Micpol replied:
Done

WrobelKrakow commented on core/src/main/java/pl/netigen/core/utils/Utils.java:
```diff
@@ -42,7 +42,7 @@ public static boolean isPermissionGranted(AppCompatActivity activity, String per
     @NonNull
     private static String getMarketLink(Activity activity, String packageName) {
-        int link_id = Config.Companion.isSamsung() ? R.string.samsung_app_id_link_netigen : R.string.play_app_id_link_netigen;
+        int link_id = Config.Companion.isSamsungStatic() ? R.string.samsung_app_id_link_netigen : R.string.play_app_id_link_netigen;
```
Why change good old name isSamsung()?

Micpol replied:
Changed it already in https://github.com/netigenkluzowicz/api_android/pull/74/commits/bd04cb75bb348e26b3e15c039dfb254f17da632f

Micpol committed:
Makes listener passed as constructor parameter usable, ads method to remove listeners

Micpol committed:
Renames isSamsungStatic method to isSamsung

Micpol committed:
Minor fixes:
-removes redundant callbacks and error check on loading
-changes scopes and names of default values

Micpol committed:
Removes isMultiScreenApp flag

Micpol committed:
Refactors code, removes redundant variables

Micpol committed:
Adds CallSuper requirement to initAdsManager, removes redundant call to hideAds

Micpol committed:
Changes noAdsSku modifier to open

Micpol committed:
Replaces onNoAdsBought implementation with hide ads

Micpol committed:
Deletes redundant method

Micpol committed:
Deletes redundant class

Micpol committed:
Removes listeners on rewardedAd destroy

WrobelKrakow commented on core/src/main/java/pl/netigen/core/ads/AdsManager.kt:
```diff
@@ -125,11 +132,24 @@ class AdsManager(var viewModel: NetigenViewModel, val activity: AppCompatActivit
+    }
+
+    fun splashScreenOnDestroy() {
+        interstitialAdManager.interstitialAdHandler.removeCallbacksAndMessages(null)
```
Here should be rather interstitialAdManager.destroy() or onDestroySplash() , interstitialAdManager.interstitialAdHandler shoudl not be public

Micpol replied:
Done

WrobelKrakow commented on core/src/main/java/pl/netigen/core/netigenapi/NetigenViewModel.kt:
```diff
@@ -1,38 +1,49 @@
+            } else {
+                getApplication<Application>().packageName + ".noads"
+            }
+        }
```
open val noAdsSku = {
        if (config.inDebugMode) {
            PaymentManager.TEST_PURCHASED
        } else {
            application.packageName + ".noads"
        }
    }

Micpol replied:
Done

Micpol committed:
Changes unnecessary var

Micpol committed:
Fixes onDestroy activity logic

Micpol committed:
Removes redundant code
